Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp

```diff
@@ -885,54 +929,65 @@ bool Storage::shouldStoreBodyAsBlob(const Data& bodyData)
+    addToRecordFilter(record.key);
 
+    backgroundIOQueue().dispatch([this, identifier, record = crossThreadCopy(record)] {
+        auto recordDirectorPath = recordDirectoryPathForKey(record.key);
+        auto recordPath = recordPathForKey(record.key);
         FileSystem::makeAllDirectories(recordDirectorPath);
 
-        ++writeOperation.activeCount;
-
-        bool shouldStoreAsBlob = shouldStoreBodyAsBlob(writeOperation.record.body);
-        auto blob = shouldStoreAsBlob ? storeBodyAsBlob(writeOperation) : std::nullopt;
+        addWriteOperationActivity(identifier);
 
-        auto recordData = encodeRecord(writeOperation.record, blob);
+        bool shouldStoreAsBlob = shouldStoreBodyAsBlob(record.body);
+        auto blob = shouldStoreAsBlob ? storeBodyAsBlob(identifier, record) : std::nullopt;
+        auto recordData = encodeRecord(record, blob);
 
         auto channel = IOChannel::open(WTFMove(recordPath), IOChannel::Type::Create);
         size_t recordSize = recordData.size();
-        channel->write(0, recordData, WorkQueue::main(), [this, &writeOperation, recordSize](int error) {
+        channel->write(0, recordData, WorkQueue::main(), [this, identifier, recordSize](int error) {
             // On error the entry still stays in the contents filter until next synchronization.
             m_approximateRecordsSize += recordSize;
-            finishWriteOperation(writeOperation, error);
+            finishWriteOperationActivity(identifier, error);
 
             LOG(NetworkCacheStorage, "(NetworkProcess) write complete error=%d", error);
         });
     });
 }
 
-void Storage::finishWriteOperation(WriteOperation& writeOperation, int error)
+void Storage::addWriteOperationActivity(WriteOperationIdentifier identifier)
 {
-    ASSERT(RunLoop::isMain());
-    ASSERT(writeOperation.activeCount);
-    ASSERT(m_activeWriteOperations.contains(&writeOperation));
+    Locker locker { m_activitiesLock };
+    m_writeOperationActivities.add(identifier);
+}
 
-    if (--writeOperation.activeCount)
-        return;
+bool Storage::removeWriteOperationActivity(WriteOperationIdentifier identifier)
+{
+    Locker locker { m_activitiesLock };
+    ASSERT(m_writeOperationActivities.contains(identifier));
+    return m_writeOperationActivities.remove(identifier);
+}
 
-    Ref protectedThis { *this };
+void Storage::finishWriteOperationActivity(WriteOperationIdentifier identifier, int error)
+{
+    ASSERT(RunLoop::isMain());
+    if (!removeWriteOperationActivity(identifier))
+        return;
 
-    if (writeOperation.completionHandler)
-        writeOperation.completionHandler(error);
+    auto* writeOperation = m_activeWriteOperations.get(identifier);
+    ASSERT(writeOperation);
+    writeOperation->invokeCompletionHandler(error);
 
-    m_activeWriteOperations.remove(&writeOperation);
+    // WriteOperation may hold last reference to this.
+    Ref protectedThis { *this };
```

Comment:
This really goes against our modern patterns. The call site should be holding a smart pointer to this object instead.